### PR TITLE
Release/3.1.0

### PR DESCRIPTION
**Pull request by CFlaniganMide**

### CODE REVIEW

**Review by StokesMIDE — APPROVED** [review]
Looks good to me!

File: examples/reading_and_writing.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
+"""
```
Comment (CrepeGoat) [inline examples/reading_and_writing.py:1]: can we add a test that runs this example? I think we could even just `pytest.monkeypatch` the `print` function and check the output directly

**Review by CrepeGoat — COMMENTED** [review]
seems good, though I'd like to change that one docstring.

the other stuff I think is not super important but does touch code/test behavior unlike a simple docstring edit, so it's probably not appropriate to address them in a release branch.

File: examples/reading_and_writing.py (hunk 1 above)
Comment (CFlaniganMide) [reply examples/reading_and_writing.py:1]: I definitely think that's a good idea, but it should probably be in a new feature branch, and won't effect release code.

File: ebmlite/util.py [hunk 2]
```diff
@@ -11,20 +11,109 @@
...
+def createID(schema, idClass, exclude=(), minId=0x81, maxId=0x1FFFFFFE, count=1):
+    """ Generate unique EBML IDs. Primarily intended for use 'offline' by
+        humans creating EBML schemata.
+
+        @param schema: The `Schema` in which the new IDs must coexist.
+        @param idClass: The EBML class of ID, one of (case-insensitive):
+            * `'a'`: Class A (1 octet, base 0x8X)
+            * `'b'`: Class B (2 octets, base 0x4000)
+            * `'c'`: Class C (3 octets, base 0x200000)
+            * `'d'`: Class D (4 octets, base 0x10000000)
```
[the remark this answers is not in the document]
Comment (CFlaniganMide) [reply ebmlite/util.py:43]: The official RFC doesn't mention it, nor do the docs on the matroska.org, but they both reference ID's by class at points.  There's an older version of the matroska docs that does reference it though.

From the [ebml spec on the matroska sourceforge](http://matroska.sourceforge.net/technical/specs/index.html):

```
bits, big-endian
1xxx xxxx                                  - Class A IDs (2^7 -1 possible values) (base 0x8X)
01xx xxxx  xxxx xxxx                       - Class B IDs (2^14-1 possible values) (base 0x4X 0xXX)
001x xxxx  xxxx xxxx  xxxx xxxx            - Class C IDs (2^21-1 possible values) (base 0x2X 0xXX 0xXX)
0001 xxxx  xxxx xxxx  xxxx xxxx  xxxx xxxx - Class D IDs (2^28-1 possible values) (base 0x1X 0xXX 0xXX 0xXX)
```

File: examples/reading_and_writing.py (hunk 1 above)
Comment (StokesMIDE) [reply examples/reading_and_writing.py:1]: Is example code usually included in the unit tests? I can't recall ever seeing this, but I suppose it makes sense. Personally, I always thought of sample code more as part of the documentation.

File: ebmlite/util.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (StokesMIDE) [reply ebmlite/util.py:43]: Even though only some versions of the documentation call them classes A-D, I think I like how it is slightly removed from just specifying the length. To me, it emphasizes that it isn't simply a matter of the number of octets; only certain ranges within each length are legal.

In any case, these particular utilities are intended for moderately advanced developer use (someone developing a schema, not just parsing a MKV or IDE). I think they will understand even if they haven't previously seen the 'classes' before.

[the remark this answers is not in the document]
Comment (CrepeGoat) [reply ebmlite/util.py:43]: I do feel better about this knowing that one version of the spec mentions it.

I'm fine with this as it is. Thanks for the discussion.

File: examples/reading_and_writing.py (hunk 1 above)
Comment (CrepeGoat) [reply examples/reading_and_writing.py:1]: it definitely *is* just the documentation. but I'd still like to know that the example works; it'd be embarrassing to have someone copy the example and have it not work.

File: ebmlite/core.py [hunk 3]
```diff
@@ -758,7 +758,13 @@ def __init__(self, stream, name=None, size=None, headers=True):
                 (if present) will not appear as a root element in the document.
                 The contents of the ``EBML`` element will always be read,
                 regardless, and stored in the Document's `info` attribute.
+            @keyword ownsStream: whether to close stream object on `close()`.
```
[the remark this answers is not in the document]
Comment (CrepeGoat) [reply ebmlite/core.py]: yeah sorry, looks good here. looking over the other comments atm

File: ebmlite/util.py [hunk 4]
```diff
@@ -11,20 +11,109 @@
...
+
+def validateID(elementId):
+    """ Verify that a number is a valid EBML element ID. A `ValueError`
+        will be raised if the element ID is invalid.
+
+        Valid ranges for the four classes of EBML ID are:
+          * A: 0x81 to 0xFE
+          * B: 0x407F to 0x7FFE
+          * C: 0x203FFF to 0x3FFFFE
+          * D: 0x101FFFFF to 0x1FFFFFFE
```
[the remark this answers is not in the document]
Comment (CrepeGoat) [reply ebmlite/util.py:83]: addressed in above discussion

File: examples/reading_and_writing.py (hunk 1 above)
Comment (CrepeGoat) [reply examples/reading_and_writing.py:1]: @CFlaniganMide agreed, this can just go in a feature branch elsewhere

**Review by CrepeGoat — APPROVED** [review]
all good!

**Pull request closed without merging**